engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java

```diff
@@ -0,0 +1,353 @@
+  @ImportStatic(LogicalLayout.class)
+  static class ReadArrayElement {
+
+    @Specialization(guards = "receiver.getUnit() == Date32")
+    static Object doDay(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException {
+      return ArrowFixedArrayDate.readDay(receiver.buffer, index);
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Date64")
+    static Object doMilliseconds(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException {
+      return ArrowFixedArrayDate.readMilliseconds(receiver.buffer, index);
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int8")
+    public static Object doByte(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      var at =
+          ArrowFixedArrayInt.adjustedIndex(receiver.buffer, receiver.unit, receiver.size, index);
+      if (receiver.buffer.isNull((int) index)) {
+        return NullValue.get();
+      }
+      return receiver.buffer.get(at);
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int16")
+    public static Object doShort(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      var at =
+          ArrowFixedArrayInt.adjustedIndex(receiver.buffer, receiver.unit, receiver.size, index);
+      if (receiver.buffer.isNull((int) index)) {
+        return NullValue.get();
+      }
+      return receiver.buffer.getShort(at);
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int32")
+    public static Object doInt(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      var at =
+          ArrowFixedArrayInt.adjustedIndex(receiver.buffer, receiver.unit, receiver.size, index);
+      if (receiver.buffer.isNull((int) index)) {
+        return NullValue.get();
+      }
+      return receiver.buffer.getInt(at);
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int64")
+    public static Object doLong(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      var at =
+          ArrowFixedArrayInt.adjustedIndex(receiver.buffer, receiver.unit, receiver.size, index);
+      if (receiver.buffer.isNull((int) index)) {
+        return NullValue.get();
+      }
+      return receiver.buffer.getLong(at);
+    }
+  }
+
+  @ExportMessage
+  @ImportStatic(LogicalLayout.class)
+  static class WriteArrayElement {
+
+    @Specialization(guards = "receiver.getUnit() == Date32")
+    static void doDay(
+        ArrowFixedSizeArrayBuilder receiver,
+        long index,
+        Object value,
+        @Cached.Shared("interop") @CachedLibrary(limit = "1") InteropLibrary iop)
+        throws UnsupportedMessageException {
+      if (receiver.sealed) {
+        throw UnsupportedMessageException.create();
```

Comment:
The [writeArrayElement message](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InteropLibrary.html#writeArrayElement-java.lang.Object-long-java.lang.Object-) can return three different `InteropException`s:
- [UnsupportedMessageException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/UnsupportedMessageException.html)
- [UnsupportedTypeException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/UnsupportedTypeException.html)
- [InvalidArrayIndexException](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InvalidArrayIndexException.html)

If you throw better interop exceptions, the Enso users may get better error message on the Enso level.

Here I suggest to report `UnsupportedTypeException` with a `hint` saying the `builder` is already sealed.

Reply:
It doesn't feel like `UnsupportedTypeException` is the right one, though. It's the problem with the receiver not with arguments.

Reply:
The goal would be to improve the error message - getting `Not_Invokable` is misleading. Maybe you can return `false` from [isArrayElementModifiable](https://github.com/enso-org/enso/pull/9150/files#diff-2ad45a930106af4f2ab4cfd869e2f5fcaf07ce06a35721835b02a96ad3e8335bR341) when _sealed_ or throw your own [truffle exception](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/exception/AbstractTruffleException.html), but `UnsupportedMessageException` certainly has different meaning than _already sealed_.

>  problem with the receiver not with arguments.

Well, the _receiver changed type_ - it used to be mutable and now it is not mutable. I don't have a problem using simplest solution if it improves the error message even if it requires a bit of mental twist.